### PR TITLE
add asserts on smart ops

### PR DESCRIPTION
**Pull request by mayeul-zama**

### CODE REVIEW

**Comment by mayeul-zama** [comment]
@slab-ci cpu_fast_test

**Comment by mayeul-zama** [comment]
@slab-ci cpu_fast_test

File: tfhe/src/shortint/engine/server_side/mul.rs [hunk 1]
```diff
@@ -179,6 +181,7 @@ impl ShortintEngine {
                 self.message_extract_assign(server_key, ct_left)?;
                 self.message_extract_assign(server_key, ct_right)?;
             }
+            assert!(server_key.is_mul_small_carry_possible(ct_left, ct_right));
```
Comment (mayeul-zama) [inline tfhe/src/shortint/engine/server_side/mul.rs:184]: Remove duplicate

**Comment by IceTDrinker** [comment]
needs a rebase to get rid of the wasm issue

**Review by IceTDrinker — COMMENTED** [review]
general question, what are the conditions that would make the smart operations fail ? i.e. a bad parameter set e.g. ?

Are the smart operations the only operations needing those checks ?

**Comment by mayeul-zama** [comment]
> general question, what are the conditions that would make the smart operations fail ? i.e. a bad parameter set e.g. ?

It should not fail if operations are used properly (except if we have bugs).
It helped catch bugs fixed in 76aaa5669198aaf7d27de294fda263bb45543dfb, 096e320b97a1cda873ff2599ad8dd1c6b332a490 and f08ea8cf85a7ca6dee85fd1bae98be0b796bd262

But for example, if you call `smart_small_scalar_mul_parallelized` with a big scalar (as before fix commit 76aaa5669198aaf7d27de294fda263bb45543dfb), it used to do a full propagate and then do the `unchecked_small_scalar_mul_parallelized` which was incorrect. Now it should panic.

> Are the smart operations the only operations needing those checks ?

IMO it could make sense for some default operations but I'm not sure.
For unchecked and checked operations, I don't think it would make sense

**Review by IceTDrinker — APPROVED** [review]

**Pull request merged**